### PR TITLE
New notebook graph functionality 

### PR DESCRIPTION
**Pull request by Batch21**
I’ve moved the notebook `draw_graph` function into a class and added a method to the class that will read the graph and save the positions of any node moved by a user into a new model json file. This hopefully deals with the frustration of spending ages moving nodes around to make a model graph readable only to lose this work when the notebook is reloaded. I’ve also added a method that will save the d3 graph into an html file, which should make it easier to share.

Still needs some work but any initial comments @pywr/pushers ?

### CODE REVIEW

**Comment by jetuk** [comment]
Tests failing because `template.html` is not packaged. It needs adding to the package date in `setup.py`.

**Comment by jetuk** [comment]
@Batch21 if you fix the tests I'll give this a review.

**Comment by Batch21** [comment]
@jetuk I've fixed the tests, which also involved moving the functions used to create the d3 graph data back out of the new schematic class. I'm having second thoughts about whether the `draw_graph` func should be in the class as it will break existing notebooks where it is used. Could have it duplicated as a function with a depreciation warning?

**Comment by jetuk** [comment]
> involved moving the functions used to create the d3 graph data back out of the new schematic class.

Why does it involve doing this?

File: pywr/notebook/__init__.py [hunk 1]
```diff
@@ -17,8 +16,110 @@
...
+
+
+class PywrSchematic:
+
+    def __init__(self, model, width=500, height=400, labels=False, attributes=False, css=None):
+        """This object contains methods that allow the graph of a pywr model network to be displayed in a jupyter
+        notebook or saved to an html file.
+
+        It also contains a method to save the node positions of a notebook graph back to a pywr model json file. Note
+        that this method currently odoes not work if the object has be instantiated using a model object.
```
Comment (jetuk) [inline pywr/notebook/__init__.py:34]: `odoes` typo

File: pywr/notebook/__init__.py [hunk 2]
```diff
@@ -17,8 +16,110 @@
...
+            If True, each graph node is labelled with its name. If false, the node names are displayed
+            during mouseover events
+        attributes : bool
+            If True, a table of node attributes is displayed during mouseover events
+        css : string
+            Stylesheet data to use instead of default
+        """
+
+        if isinstance(model, Model):
+            self.graph = self.pywr_model_to_d3_json(model, attributes)
```
Comment (jetuk) [inline pywr/notebook/__init__.py:54]: Call to a method that doesn't exist? Should be the function instead?

File: pywr/notebook/__init__.py [hunk 3]
```diff
@@ -17,8 +16,110 @@
...
+            # TODO update when schema branch is merged
+            self.json = None
+        else:
+            self.graph = self.pywr_json_to_d3_json(model, attributes)
```
Comment (jetuk) [inline pywr/notebook/__init__.py:58]: Call to a method that doesn't exist? Should be the function instead?

File: pywr/notebook/__init__.py [hunk 4]
```diff
@@ -17,8 +16,110 @@
...
+                model_data=json.dumps(self.json),
+                height=self.height,
+                width=self.width,
+                filename=json.dumps(filename)
+            )))
+
+    def to_html(self, filename, title="Model Schematic"):
+        """Save an HTML file of schematic"""
+
+        # TODO add option to get node position from graph that has already been drawn in a notebook
```
Comment (jetuk) [inline pywr/notebook/__init__.py:102]: What would this look like? Is it hard?

File: pywr/notebook/__init__.py [hunk 5]
```diff
@@ -106,8 +207,8 @@ def get_node_attr(node):
         attr_type = type(attr_val).__name__
 
         attrs_to_skip = ["component_attrs", "components", "color", "model",  "input", "output",
-                         "inputs", "outputs", "sub_domain", "sub_output", "sublinks", "visible",
-                         "fully_qualified_name", "allow_isolated"]
+                        "inputs", "outputs", "sub_domain", "sub_output", "sublinks", "visible",
```
Comment (jetuk) [inline pywr/notebook/__init__.py:210]: Why delete these spaces? PEP8 violation!

File: pywr/notebook/__init__.py [hunk 6]
```diff
@@ -17,8 +16,110 @@
...
+        labels : bool
+            If True, each graph node is labelled with its name. If false, the node names are displayed
+            during mouseover events
+        attributes : bool
+            If True, a table of node attributes is displayed during mouseover events
+        css : string
+            Stylesheet data to use instead of default
+        """
+
+        if isinstance(model, Model):
```
Comment (jetuk) [inline pywr/notebook/__init__.py:53]: You could instead have a classmethod for each of these functions. E.g. `from_json` and `from_model`. Or just make them static methods. I think most notebooks use `draw_graph` which has been broken here anyway?

File: pywr/notebook/__init__.py [hunk 7]
```diff
@@ -212,50 +313,3 @@ def create_node_class_trees():
         clss = [cls.__name__.lower() for cls in classes[::-1]]
         node_class_trees[name] = clss
     return node_class_trees
-
-
-def draw_graph(model, width=500, height=400, labels=False, attributes=False, css=None):
```
Comment (jetuk) [inline pywr/notebook/__init__.py:217]: We could replace this with a function that creates a schematic instance and calls `draw_graph`?

File: pywr/notebook/draw_graph.js [hunk 8]
```diff
@@ -136,6 +137,12 @@ require(["d3"], function(d3) {
 
     function tick() {
 
+        node.attr("transform", function(d) {
```
Comment (jetuk) [inline pywr/notebook/draw_graph.js:140]: Add a comment that this is bounding the node positions to stay inside the box. 

An alternative to this is to retain the coordinates as is, but use a `viewBox` on the SVG element to view the box containing the model?

File: pywr/notebook/save_graph.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+require(["d3"], function(d3) {
+
+    const model_data = {{ model_data }};
+
+    const width = {{ width }},
+        height = {{ height }};
+
+    const nodes = d3.select(".pywr_schematic").selectAll(".node").data();
+
+    const node_data_obj = {}
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:14]: ;

File: pywr/notebook/save_graph.js [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    const posX = d3.scaleLinear()
+                    .range([-100, 100])
+                    .domain([0, width]);
+    const posY = d3.scaleLinear() 
+                   .range([100, -100])
+                   .domain([0, height]);
+    
+    for (let i = 0; i < model_data.nodes.length; i++){
+        let node_name = model_data.nodes[i].name;
+        let node_data = node_data_obj[node_name]
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:29]: ;

File: pywr/notebook/__init__.py [hunk 11]
```diff
@@ -17,8 +16,110 @@
...
+    def to_html(self, filename, title="Model Schematic"):
+        """Save an HTML file of schematic"""
+
+        # TODO add option to get node position from graph that has already been drawn in a notebook
+
+        js = draw_graph_template.render(
+            graph=self.graph,
+            width=self.width,
+            height=self.height,
+            element=json.dumps(".schematic"),
```
Comment (jetuk) [inline pywr/notebook/__init__.py:108]: Why do you need to `json.dumps` this string?

File: pywr/notebook/save_graph.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        if (node_data.fixed){
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:30]: Worth having an option to save all node positions even if not fixed?

File: pywr/notebook/save_graph.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+            model_data["nodes"][i]["position"] = {"schematic": [posX(node_data.x), posY(node_data.y)]}
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:31]: This will nuke any existing position information on the node (e.g. "geographic" coordinates.)

File: pywr/notebook/save_graph.js [hunk 14]
```diff
@@ -0,0 +1,53 @@
+
+
+require.config({paths: {d3: 'https://d3js.org/d3.v5.min'}});
+  
+require(["d3"], function(d3) {
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:5]: Add a comment / docstring

**Comment by jetuk** [comment]
@Batch21 could you look in to #870 and merge if OK, then rebase this against the new master?

**Comment by jetuk** [comment]
@Batch21 is this ready now?

File: pywr/notebook/__init__.py (hunk 11 above)
Comment (Batch21) [reply pywr/notebook/__init__.py:124]: Without it, the string does not get properly rendered with `"` in the js file

**Comment by Batch21** [comment]
@jetuk still need to respond to a few of the comment and tidy up

File: pywr/notebook/save_graph.js (hunk 14 above)
Comment (Batch21) [reply pywr/notebook/save_graph.js:5]: Not sure how to format this as the function is an anonymous one passed as a callback to require. The alternative could be to add some more comments to the code within the function?

File: pywr/notebook/__init__.py (hunk 7 above)
Comment (Batch21) [reply pywr/notebook/__init__.py:217]: Done

File: pywr/notebook/__init__.py (hunk 6 above)
Comment (Batch21) [reply pywr/notebook/__init__.py:53]: Do you mean converting the `to_d3` functions into static or class methods rather than standalone functions?

File: pywr/notebook/__init__.py (hunk 4 above)
Comment (Batch21) [reply pywr/notebook/__init__.py:126]: I can't think of a way of doing it without making the existing `draw_graph.js` script even messier - might be better to have a separate script.

I think I might leave this and see if the method get used at all

**Comment by Batch21** [comment]
@jetuk  - I've addressed most of your comments and asked for clarification on a few

File: pywr/notebook/__init__.py [hunk 15]
```diff
@@ -82,38 +82,46 @@ def draw_graph(self):
...
 
-    def save_positions(self, filename="node_positions.csv"):
-        """Save a copy of the model json with update schematic positions"""
+        Parameters
+        ----------
+        filename: str
+            The name of the file to save the output data to.
+        save_unfixed: bool
+            If True, then all node position are saved to output file. If False, only nodes who have had their position
+            fixed in the d3 grpah have their positions saved.
```
Comment (jetuk) [inline pywr/notebook/__init__.py:96]: `grpah`

File: pywr/notebook/__init__.py [hunk 16]
```diff
@@ -82,38 +82,46 @@ def draw_graph(self):
...
+        filetype: str
+            Should be either 'json' to save the model data with updated node positions to a json file or 'csv' to save
+            node positions to a csv file.
+        """
 
-        if self.json is None:
-            warnings.warn("Node positions cannot be saved to json if PywrSchematic object has been instantiated using"
-                          "a pywr model object. Please use a json file path or Python dict instead", stacklevel=2)
+        if self.json is None and filetype == "json":
+            warnings.warn("Node positions cannot be saved to json if PywrSchematic object has been instantiated using "
```
Comment (jetuk) [inline pywr/notebook/__init__.py:103]: `json` -> `JSON` in docstring

File: pywr/notebook/save_graph.js [hunk 17]
```diff
@@ -4,54 +4,59 @@ require.config({paths: {d3: 'https://d3js.org/d3.v5.min'}});
...
                 node_data.fixed,
                 posX(node_data.x),
                 posY(node_data.y),
-            ]
-            output_data.push(position_data.join(","))
+            ];
+            output_data.push(position_data.join(","));
         }
-        download(filename, output_data.join("\n"));
+        // TODO append error message if filetype not recognised
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:52]: Yes, I guess this should raise an error, but what is the JS equivalent of that?

File: pywr/notebook/save_graph.js [hunk 18]
```diff
@@ -4,54 +4,59 @@ require.config({paths: {d3: 'https://d3js.org/d3.v5.min'}});
...
     }
 
-    
+    if (filetype == "json"){
+        download(filename, JSON.stringify(model_data));
```
Comment (jetuk) [inline pywr/notebook/save_graph.js:56]: Why did you move these calls to download to a separate if-else-if statement and not keep them in the above blocks?

File: pywr/notebook/save_graph.js (hunk 17 above)
Comment (Batch21) [reply pywr/notebook/save_graph.js]: I added a check for this in python instead

File: pywr/notebook/save_graph.js (hunk 18 above)
Comment (Batch21) [reply pywr/notebook/save_graph.js:55]: The check above is in a for loop which is why it is redone here

**Comment by jetuk** [comment]
Thanks @Batch21 looks good, I'll merge this now. Hopefully we can give it a bit of road test before the next release. It would be good to think what to do with this good in the long term. We don't test it very well at the moment.
